File: day-05/python/day5.py
#!/usr/bin/env python3

import sys

def part1(d):
    return d[-1]

def part2(d):
    for x in range(d[0], d[-1]):
        if x not in d:
            return x

data = [x.replace('F', '0').replace('B', '1').replace('L', '0').replace('R', '1') for x in open(sys.argv[1], "r").read().split("\n")]

sID = []

for line in data:
    fRow = int(line[:7], 2)
    fCol = int(line[-3:], 2)

    sID.append(fRow * 8 + fCol)

sID.sort()

print(f"Part 1: {part1(sID)}\nPart 2: {part2(sID)}")

File: day-05/nim/day5.nim
from os import paramStr
import strutils, strformat, sugar, algorithm

let data = collect(newSeq):
  for x in lines(paramStr(1)):
    x.multiReplace(("F", "0"), ("B", "1"), ("L", "0"), ("R", "1"))
var sID = newSeq[int](0)

proc part1(d: seq[int]): int =
  return d[^1]

proc part2(d: seq[int]): int =
  for x in d[0]..d[^1]:
    if x notin d:
      return x

for line in data:
  var
    fRow = parseBinInt(line[0..^4])
    fCol = parseBinInt(line[^3..^1])

  sID.add(fRow * 8 + fCol)

sort(sID)

echo fmt"Part 1: {part1(sID)}", "\n", fmt"Part 2: {part2(sID)}"
